### ISSUE TITLE
Exception in UnitThrowAnalysis

### ISSUE THREAD

**Issue by bergerbd** [Sable/soot#358]
Soot has some problems to jimplify a method. It seems that the type of the processed throw statement is bottom_type instead of an actual type. The error occurs if I load one of the following methods:

```
<android.support.v4.print.PrintHelperKitkat: android.graphics.Bitmap loadConstrainedBitmap(android.net.Uri,int)>
<android.support.v4.app.NotificationManagerCompat: java.util.Set getEnabledListenerPackages(android.content.Context)>
```
I think Soot is using sdk level 14 (since it is the min sdk according to the manifest). 

Version: HEAD
Stacktrace:
```java
java.lang.IllegalStateException: UnitThrowAnalysis StmtSwitch: type of throw argument is not a RefType!
	at soot.toolkits.exceptions.AbstractThrowAnalysis.mightThrowExplicitly(AbstractThrowAnalysis.java:65) ~[sootclasses.jar:trunk]
	at soot.toolkits.exceptions.UnitThrowAnalysis$UnitSwitch.caseThrowStmt(UnitThrowAnalysis.java:706) ~[sootclasses.jar:trunk]
	at soot.jimple.internal.JThrowStmt.apply(JThrowStmt.java:70) ~[sootclasses.jar:trunk]
	at soot.toolkits.exceptions.UnitThrowAnalysis.mightThrow(UnitThrowAnalysis.java:241) ~[sootclasses.jar:trunk]
	at soot.toolkits.graph.ExceptionalUnitGraph.getExceptionDests(ExceptionalUnitGraph.java:800) ~[sootclasses.jar:trunk]
	at soot.toolkits.graph.ExceptionalUnitGraph.buildHeadsAndTails(ExceptionalUnitGraph.java:763) ~[sootclasses.jar:trunk]
	at soot.toolkits.graph.ExceptionalUnitGraph.initialize(ExceptionalUnitGraph.java:293) ~[sootclasses.jar:trunk]
	at soot.toolkits.graph.ExceptionalUnitGraph.<init>(ExceptionalUnitGraph.java:159) ~[sootclasses.jar:trunk]
	at soot.toolkits.exceptions.TrapTightener.internalTransform(TrapTightener.java:79) ~[sootclasses.jar:trunk]
	at soot.BodyTransformer.transform(BodyTransformer.java:51) ~[sootclasses.jar:trunk]
	at soot.BodyTransformer.transform(BodyTransformer.java:58) ~[sootclasses.jar:trunk]
	at soot.BodyTransformer.transform(BodyTransformer.java:63) ~[sootclasses.jar:trunk]
	at soot.dexpler.DexBody.jimplify(DexBody.java:625) ~[sootclasses.jar:trunk]
	at soot.dexpler.DexMethod$1.getBody(DexMethod.java:150) ~[sootclasses.jar:trunk]
	at soot.SootMethod.getBodyFromMethodSource(SootMethod.java:91) ~[sootclasses.jar:trunk]
	at soot.SootMethod.retrieveActiveBody(SootMethod.java:324) ~[sootclasses.jar:trunk]
... my code ...
```

**Comment by bergerbd** [comment]
There has been a similar report #312

**Comment by bergerbd** [comment]
The problematic statement is a throw statement that throws a local that is not set ($u6). Since the code is dead it could be eliminated. I've no idea if this is a bug of some earlier steps during jimplification.

```
    private android.graphics.Bitmap loadConstrainedBitmap(android.net.Uri, int) throws java.io.FileNotFoundException
    {
        unknown $u9, $u10, $u11, $u0, $u1, $u2, $u3, $u4, $u5, $u6, $u7, $u-1, $u6#2, $u-1#2, $u7#4, $u7#5, $u6#3, $u6#4, $u6#5, $u-1#3, $u7#6, $u7#7;
        java.lang.Throwable $u6#7, $u6#8;

        $u9 := @this: android.support.v4.print.PrintHelperKitkat;

        $u10 := @parameter0: android.net.Uri;

        $u11 := @parameter1: int;

        if $u11 <= 0 goto label01;

        if $u10 == null goto label01;

        $u7 = $u9.<android.support.v4.print.PrintHelperKitkat: android.content.Context mContext>;

        if $u7 != null goto label02;

     label01:
        $u6#2 = new java.lang.IllegalArgumentException;

        specialinvoke $u6#2.<java.lang.IllegalArgumentException: void <init>(java.lang.String)>("bad argument to getScaledBitmap");

        throw $u6#2;

     label02:
        $u3 = new android.graphics.BitmapFactory$Options;

        specialinvoke $u3.<android.graphics.BitmapFactory$Options: void <init>()>();

        $u3.<android.graphics.BitmapFactory$Options: boolean inJustDecodeBounds> = 1;

        specialinvoke $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.Bitmap loadBitmap(android.net.Uri,android.graphics.BitmapFactory$Options)>($u10, $u3);

        $u5 = $u3.<android.graphics.BitmapFactory$Options: int outWidth>;

        $u1 = $u3.<android.graphics.BitmapFactory$Options: int outHeight>;

        if $u5 <= 0 goto label13;

        if $u1 > 0 goto label03;

        return null;

     label03:
        $u-1 = staticinvoke <java.lang.Math: int max(int,int)>($u5, $u1);

        $u2 = $u-1;

        $u4 = 1;

     label04:
        if $u2 <= $u11 goto label05;

        $u2 = $u2 >>> 1;

        $u4 = $u4 << 1;

        goto label04;

     label05:
        if $u4 <= 0 goto label14;

        $u-1#2 = staticinvoke <java.lang.Math: int min(int,int)>($u5, $u1);

        $u7#4 = $u-1#2 / $u4;

        if $u7#4 <= 0 goto label15;

        $u7#5 = $u9.<android.support.v4.print.PrintHelperKitkat: java.lang.Object mLock>;

        entermonitor $u7#5;

     label06:
        $u6#3 = new android.graphics.BitmapFactory$Options;

        specialinvoke $u6#3.<android.graphics.BitmapFactory$Options: void <init>()>();

        $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.BitmapFactory$Options mDecodeOptions> = $u6#3;

        $u6#4 = $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.BitmapFactory$Options mDecodeOptions>;

        $u6#4.<android.graphics.BitmapFactory$Options: boolean inMutable> = 1;

        $u6#5 = $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.BitmapFactory$Options mDecodeOptions>;

        $u6#5.<android.graphics.BitmapFactory$Options: int inSampleSize> = $u4;

        $u0 = $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.BitmapFactory$Options mDecodeOptions>;

        exitmonitor $u7#5;

     label07:
        $u-1#3 = specialinvoke $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.Bitmap loadBitmap(android.net.Uri,android.graphics.BitmapFactory$Options)>($u10, $u0);

     label08:
        $u7#6 = $u9.<android.support.v4.print.PrintHelperKitkat: java.lang.Object mLock>;

        entermonitor $u7#6;

        $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.BitmapFactory$Options mDecodeOptions> = null;

        exitmonitor $u7#6;

        return $u-1#3;

     label09:
        $u6#7 := @caughtexception;

     label10:
        exitmonitor $u7#5;

     label11:
        throw $u6#7;

     label12:
        $u6#8 := @caughtexception;

        $u7#7 = $u9.<android.support.v4.print.PrintHelperKitkat: java.lang.Object mLock>;

        entermonitor $u7#7;

        $u9.<android.support.v4.print.PrintHelperKitkat: android.graphics.BitmapFactory$Options mDecodeOptions> = null;

        exitmonitor $u7#7;

        throw $u6#8;

        throw $u6;

     label13:
        return null;

     label14:
        return null;

     label15:
        return null;

        catch java.lang.Throwable from label06 to label07 with label09;
        catch java.lang.Throwable from label07 to label08 with label12;
        catch java.lang.Throwable from label10 to label11 with label09;
    }
```

**Comment by StevenArzt** [comment]
Normally, Soot should already run a dead code eliminator when reading in dex files. If this problem still exists with the newest nightly build of Soot, please send me the APK file to Steven.Arzt@cased.de

**Comment by bergerbd** [comment]
Thank you very much Steven.

Fixed in 86b0e6ef590fe0f33be403df049a6f168ba02b4d . -> Closing Issue.
